File: src/Z.Caching/System.Runtime.Caching.MemoryCache/MemoryCache.AddOrGetExisting.cs
// Description: C# Extension Methods Library to enhances the .NET Framework by adding hundreds of new methods. It drastically increases developers productivity and code readability. Support C# and VB.NET
// Website & Documentation: https://github.com/zzzprojects/Z.ExtensionMethods
// Forum: https://github.com/zzzprojects/Z.ExtensionMethods/issues
// License: https://github.com/zzzprojects/Z.ExtensionMethods/blob/master/LICENSE
// More projects: http://www.zzzprojects.com/
// Copyright © ZZZ Projects Inc. 2014 - 2016. All rights reserved.
using System;
using System.Runtime.Caching;

public static partial class Extensions
{
    /// <summary>A MemoryCache extension method that adds an or get existing.</summary>
    /// <typeparam name="TValue">Type of the value.</typeparam>
    /// <param name="cache">The cache to act on.</param>
    /// <param name="key">The key.</param>
    /// <param name="value">The value.</param>
    /// <returns>A TValue.</returns>
    public static TValue AddOrGetExisting<TValue>(this MemoryCache cache, string key, TValue value)
    {
        object item = cache.AddOrGetExisting(key, value, new CacheItemPolicy()) ?? value;

        return (TValue) item;
    }

    /// <summary>A MemoryCache extension method that adds an or get existing.</summary>
    /// <typeparam name="TValue">Type of the value.</typeparam>
    /// <param name="cache">The cache to act on.</param>
    /// <param name="key">The key.</param>
    /// <param name="valueFactory">The value factory.</param>
    /// <returns>A TValue.</returns>
    public static TValue AddOrGetExisting<TValue>(this MemoryCache cache, string key, Func<string, TValue> valueFactory)
    {
        var lazy = new Lazy<TValue>(() => valueFactory(key));

        Lazy<TValue> item = (Lazy<TValue>) cache.AddOrGetExisting(key, lazy, new CacheItemPolicy()) ?? lazy;

        return item.Value;
    }

    /// <summary>A MemoryCache extension method that adds an or get existing.</summary>
    /// <typeparam name="TValue">Type of the value.</typeparam>
    /// <param name="cache">The cache to act on.</param>
    /// <param name="key">The key.</param>
    /// <param name="valueFactory">The value factory.</param>
    /// <param name="policy">The policy.</param>
    /// <param name="regionName">(Optional) name of the region.</param>
    /// <returns>A TValue.</returns>
    public static TValue AddOrGetExisting<TValue>(this MemoryCache cache, string key, Func<string, TValue> valueFactory, CacheItemPolicy policy, string regionName = null)
    {
        var lazy = new Lazy<TValue>(() => valueFactory(key));

        Lazy<TValue> item = (Lazy<TValue>) cache.AddOrGetExisting(key, lazy, policy, regionName) ?? lazy;

        return item.Value;
    }

    /// <summary>A MemoryCache extension method that adds an or get existing.</summary>
    /// <typeparam name="TValue">Type of the value.</typeparam>
    /// <param name="cache">The cache to act on.</param>
    /// <param name="key">The key.</param>
    /// <param name="valueFactory">The value factory.</param>
    /// <param name="absoluteExpiration">The policy.</param>
    /// <param name="regionName">(Optional) name of the region.</param>
    /// <returns>A TValue.</returns>
    public static TValue AddOrGetExisting<TValue>(this MemoryCache cache, string key, Func<string, TValue> valueFactory, DateTimeOffset absoluteExpiration, string regionName = null)
    {
        var lazy = new Lazy<TValue>(() => valueFactory(key));

        Lazy<TValue> item = (Lazy<TValue>) cache.AddOrGetExisting(key, lazy, absoluteExpiration, regionName) ?? lazy;

        return item.Value;
    }
}

File: src (VB.NET)/Z.ExtensionMethods.VB/Z.Caching/System.Runtime.Caching.MemoryCache/MemoryCache.AddOrGetExisting.vb

' Copyright (c) 2015 ZZZ Projects. All rights reserved
' Licensed under MIT License (MIT) (https://github.com/zzzprojects/Z.ExtensionMethods)
' Website: http://www.zzzprojects.com/
' Feedback / Feature Requests / Issues : http://zzzprojects.uservoice.com/forums/283927
' All ZZZ Projects products: Entity Framework Extensions / Bulk Operations / Extension Methods /Icon Library

Imports System.Runtime.Caching

Public Module Extensions_1

    ''' <summary>A MemoryCache extension method that adds an or get existing.</summary>
    ''' <typeparam name="TValue">Type of the value.</typeparam>
    ''' <param name="cache">The cache to act on.</param>
    ''' <param name="key">The key.</param>
    ''' <param name="value">The value.</param>
    ''' <returns>A TValue.</returns>
    <System.Runtime.CompilerServices.Extension> _
    Public Function AddOrGetExisting(Of TValue)(cache As MemoryCache, key As String, value As TValue) As TValue
        Dim item As Object = If(cache.AddOrGetExisting(key, value, New CacheItemPolicy()), value)

        Return DirectCast(item, TValue)
    End Function

    ''' <summary>A MemoryCache extension method that adds an or get existing.</summary>
    ''' <typeparam name="TValue">Type of the value.</typeparam>
    ''' <param name="cache">The cache to act on.</param>
    ''' <param name="key">The key.</param>
    ''' <param name="valueFactory">The value factory.</param>
    ''' <returns>A TValue.</returns>
    <System.Runtime.CompilerServices.Extension> _
    Public Function AddOrGetExisting(Of TValue)(cache As MemoryCache, key As String, valueFactory As Func(Of String, TValue)) As TValue
        Dim lazy = New Lazy(Of TValue)(Function() valueFactory(key))

        Dim item As Lazy(Of TValue) = If(DirectCast(cache.AddOrGetExisting(key, lazy, New CacheItemPolicy()), Lazy(Of TValue)), lazy)

        Return item.Value
    End Function

    ''' <summary>A MemoryCache extension method that adds an or get existing.</summary>
    ''' <typeparam name="TValue">Type of the value.</typeparam>
    ''' <param name="cache">The cache to act on.</param>
    ''' <param name="key">The key.</param>
    ''' <param name="valueFactory">The value factory.</param>
    ''' <param name="policy">The policy.</param>
    ''' <param name="regionName">(Optional) name of the region.</param>
    ''' <returns>A TValue.</returns>
    <System.Runtime.CompilerServices.Extension> _
    Public Function AddOrGetExisting(Of TValue)(cache As MemoryCache, key As String, valueFactory As Func(Of String, TValue), policy As CacheItemPolicy, Optional regionName As String = Nothing) As TValue
        Dim lazy = New Lazy(Of TValue)(Function() valueFactory(key))

        Dim item As Lazy(Of TValue) = If(DirectCast(cache.AddOrGetExisting(key, lazy, policy, regionName), Lazy(Of TValue)), lazy)

        Return item.Value
    End Function

    ''' <summary>A MemoryCache extension method that adds an or get existing.</summary>
    ''' <typeparam name="TValue">Type of the value.</typeparam>
    ''' <param name="cache">The cache to act on.</param>
    ''' <param name="key">The key.</param>
    ''' <param name="valueFactory">The value factory.</param>
    ''' <param name="absoluteExpiration">The policy.</param>
    ''' <param name="regionName">(Optional) name of the region.</param>
    ''' <returns>A TValue.</returns>
    <System.Runtime.CompilerServices.Extension> _
    Public Function AddOrGetExisting(Of TValue)(cache As MemoryCache, key As String, valueFactory As Func(Of String, TValue), absoluteExpiration As DateTimeOffset, Optional regionName As String = Nothing) As TValue
        Dim lazy = New Lazy(Of TValue)(Function() valueFactory(key))

        Dim item As Lazy(Of TValue) = If(DirectCast(cache.AddOrGetExisting(key, lazy, absoluteExpiration, regionName), Lazy(Of TValue)), lazy)

        Return item.Value
    End Function
End Module


